pkg/cli/cli.go

```diff
@@ -153,6 +192,40 @@ func (c *cli) initialize() error {
 		return err
 	}
 
+	// With or without a config, a user can:
+	// 1. Not set --plugins
+	// 2. Set --plugins to a plugin, ex. --plugins=go
+	// In case 1, either default plugins or the config layout can be used to
+	// determine which plugin to use. Otherwise, the value passed to --plugins
+	// is used. Since both layout and a --plugins value are user input, they
+	// may need to be resolved to known plugin keys.
+	var filterKeys map[string]string
+	switch {
+	case len(c.cliPluginKeys) != 0:
+		// Filter plugins by keys passed in CLI.
+		filterKeys = c.cliPluginKeys
+	case c.configured && !projectConfig.IsV1():
+		// All non-v1 configs must have a layout key. This check will help with
+		// migration.
+		if projectConfig.Layout == "" {
+			return fmt.Errorf("config must have a layout value")
+		}
+		// Use defaults if no config is found, otherwise infer plugins from the
+		// config's layout.
+		name, version := plugin.SplitKey(projectConfig.Layout)
+		filterKeys = map[string]string{name: version}
+	default:
+		// Use the default plugins for this project version.
+		c.resolvedPlugins = c.defaultPluginsFromOptions[c.projectVersion]
+	}
+	if len(filterKeys) != 0 {
+		plugins := c.pluginsFromOptions[c.projectVersion]
+		c.resolvedPlugins, err = resolvePluginsByKeys(plugins, filterKeys)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
what happens if filterKeys is not set here? (i.e. if there's no plugins set for options)?

Reply:
(if we check elsewhere, add a comment here)

Reply:
(it's also maybe a good indication that we're failing at the principle of "parse instead of validate")

Reply:
`cli.resolvedPlugins` defaults to the [default plugins](https://github.com/kubernetes-sigs/kubebuilder/pull/1403/files#diff-c11db1404a63b574b13465c6a00e4ee1R219) for a project version if `filterKeys` is empty. Default plugins are ensured to exist by validating them [here](https://github.com/kubernetes-sigs/kubebuilder/pull/1403/files#diff-c11db1404a63b574b13465c6a00e4ee1R305). Does this still fail parse instead of validate")? I'll add a comment describing this scenario if not.

Looks like the case when `c.configured && projectConfig.IsV1() == true` is not validated however. Making a follow-up now.